presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipTransformFunction.java

```diff
@@ -0,0 +1,137 @@
+import static java.lang.Math.min;
+
+public final class ZipTransformFunction
+        extends SqlScalarFunction
+{
+    public static final ZipTransformFunction ZIP_TRANSFORM_FUNCTION = new ZipTransformFunction();
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ZipTransformFunction.class, "transform", Type.class, Type.class, Type.class, Block.class, Block.class, MethodHandle.class);
+
+    private ZipTransformFunction()
+    {
+        super(new Signature(
+                "zip",
+                FunctionKind.SCALAR,
+                ImmutableList.of(typeVariable("T"), typeVariable("U"), typeVariable("V")),
+                ImmutableList.of(),
+                parseTypeSignature("array(V)"),
+                ImmutableList.of(parseTypeSignature("array(T)"), parseTypeSignature("array(U)"), parseTypeSignature("function(T,U,V)")),
+                false));
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return false;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "merge two arrays, element-wise, into a single array of rows, and apply lambda to each row";
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type leftArrayType = boundVariables.getTypeVariable("T");
+        Type rightArrayType = boundVariables.getTypeVariable("U");
+        Type outputType = boundVariables.getTypeVariable("V");
+        return new ScalarFunctionImplementation(
+                false,
+                ImmutableList.of(false, false, false),
+                METHOD_HANDLE.bindTo(leftArrayType).bindTo(rightArrayType).bindTo(outputType),
+                isDeterministic());
+    }
+
+    public static Block transform(Type leftArrayType, Type rightArrayType, Type outputType, Block leftBlock, Block rightBlock, MethodHandle function)
+    {
+        int leftPositionCount = leftBlock.getPositionCount();
+        int rightPositionCount = rightBlock.getPositionCount();
+        BlockBuilder resultBuilder = outputType.createBlockBuilder(new BlockBuilderStatus(), max(leftPositionCount, rightPositionCount));
+        int shortest = min(leftPositionCount, rightPositionCount);
+        for (int position = 0; position < shortest; position++) {
+            Object left = readNativeValue(leftArrayType, leftBlock, position);
+            Object right = readNativeValue(rightArrayType, rightBlock, position);
+            Object output;
+            try {
+                output = function.invoke(left, right);
+            }
+            catch (Throwable throwable) {
+                throw Throwables.propagate(throwable);
+            }
+            writeNativeValue(outputType, resultBuilder, output);
+        }
+        if (leftPositionCount > shortest) {
```

Comment:
IIRC, we discusses and agreed that the function will fail if the two arrays are not of the same length.